mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp

```diff
@@ -82,22 +90,43 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
         rewriter.setInsertionPointAfter(callee);
         unallocFn =
             rewriter.create<func::FuncOp>(loc, fnName, fnType, visibility, nullptr, nullptr);
-        // clone the body.
+        // Clone the body.
+        // TODO: we do not support submodules, i.e. our gradient functions
+        // will have just one block.
+        assert(callee.getBody().hasOneBlock() &&
+               "Gradients with quantum submodules are not supported");
```

Comment:
subroutines? 😅 
```suggestion
        // TODO: we do not support subroutines, i.e. our gradient functions
        // will have just one block.
        assert(callee.getBody().hasOneBlock() &&
               "Gradients with quantum subroutines are not supported");
```

Although this is odd, if you want to check that there are no subroutines, you should verify that there are no call ops (or call-like ops) in the body. Checking for a single block just asserts that there is no unstructured control flow.

Reply:
Oh right, I remember now. I think you wanted to check that there is only a single block so that we can assume there is only one return statement? Which I think is true anyway, as we only deal with structured control flow until going to llvm, so that makes sense to assert.